### ISSUE TITLE
Fetching SSM params: InvalidSignatureException: Signature expired

### ISSUE THREAD

**Issue by hamilton-s** [middyjs/middy#1123]
**Describe the bug**
Since the end of October, we have seen our Lambda functions intermittently fail due to SSM parameters not being fetched. The error we are seeing looks like the following:

```
InvalidSignatureException: Signature expired: 20231103T171116Z is now earlier than 20231103T171224Z (20231103T171724Z - 5 min.)
 at throwDefaultError (/var/runtime/node_modules/@aws-sdk/smithy-client/dist-cjs/default-error-handler.js:8:22)
 at /var/runtime/node_modules/@aws-sdk/smithy-client/dist-cjs/default-error-handler.js:18:39
 at de_GetParametersCommandError (/var/runtime/node_modules/@aws-sdk/client-ssm/dist-cjs/protocols/Aws_json1_1.js:4194:20)
 at process.processTicksAndRejections ...
```

We have managed to bring our errors down by disabling prefetch and reducing the cacheExpiry - however, we would prefer to keep these options as they were. We're also not sure why the issue has suddenly started happening as the issues didn't seem to coincide with a particular upgrade/code change.

**To Reproduce**
```
export default middy(handler)
  .use(
    ssm({
      fetchData: {
        paramA: `some/path/a`,
        paramB: `some/path/b`,
      },
      cacheExpiry: 600000,
      setToContext: true,
    })
  )
```
We notice it only fails in around 1-2% of cases in our production environments.

**Environments**
 - Node.js: 18
 - "@middy/ssm": "4.6.5"
 - "@middy/core": "4.6.5"
 - "@aws-sdk/client-lambda": "3.425.0"

**Additional context**
We've noticed this across a range of our services, with different versions of @middy/ssm

**Comment by willfarrell** [comment]
Thanks for reporting. As you mentioned "We're also not sure why the issue has suddenly started happening as the issues didn't seem to coincide with a particular upgrade/code change.", which makes me think there was a change on the AWS infra side, but I don't see anything in the docs that jumps out as changed. I wonder if there was an SDK change?

If you reached out the AWS support, can you shared their response here.

I'll do some digging as well.

**Comment by willfarrell** [comment]
For what you described, that sounds like it could easily cause this issue. I'll close for now. If you you need to reopen, please do so. Other are welcome to comment if they're also running into this.

**Comment by HumbleBeck** [comment]
Hello,
I've hit the same error on Lambda@Edge recently. 
Setup looks like
```javascript
export default middy(handler)
  .use(doNotWaitForEmptyEventLoop())
  .use(ssm({
    fetchData: {
      paramA: "path"
    },
    setToContext: true,
    awsClientOptions: {
      region: process.env.REGION || 'us-east-1'
    }
  }))
```
Environment
- Node.js: 18
- "@middy/ssm": "4.6.5"
- "@middy/core": "4.6.5"

```
INFO	InvalidSignatureException: Signature expired: 20231119T162823Z is now earlier than 20231119T163124Z (20231119T163624Z - 5 min.)
    at throwDefaultError (/var/runtime/node_modules/@aws-sdk/smithy-client/dist-cjs/default-error-handler.js:8:22)
    at /var/runtime/node_modules/@aws-sdk/smithy-client/dist-cjs/default-error-handler.js:18:39
    at de_GetParametersByPathCommandError (/var/runtime/node_modules/@aws-sdk/client-ssm/dist-cjs/protocols/Aws_json1_1.js:4242:20)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /var/runtime/node_modules/@aws-sdk/middleware-serde/dist-cjs/deserializerMiddleware.js:7:24
    at async /var/runtime/node_modules/@aws-sdk/middleware-signing/dist-cjs/awsAuthMiddleware.js:14:20
    at async /var/runtime/node_modules/@aws-sdk/middleware-retry/dist-cjs/retryMiddleware.js:27:46
    at async /var/runtime/node_modules/@aws-sdk/middleware-logger/dist-cjs/loggerMiddleware.js:7:26
    at async Promise.allSettled (index 0)
    at async to (/var/task/src/edgeGate/handler.js:63:43001) {
  '$fault': 'client',
  '$metadata': {
    httpStatusCode: 400,
    requestId: '55d1f065-3954-4531-9ffe-af7e2e2a8a29',
    extendedRequestId: undefined,
    cfId: undefined,
    attempts: 1,
    totalRetryDelay: 0
  },
  __type: 'InvalidSignatureException'
}
```

It happens randomly to a small number of requests. I couldn't figure out a pattern yet.

**Comment by hamilton-s** [comment]
We managed to fix our observability lambda layer issue but we are still experiencing the middy issue.  @willfarrell Can we please reopen this issue since others appear to have the same issue in the thread above too?

**Comment by HumbleBeck** [comment]
My theory on what is happening in our case:
- here we cache a client in memory https://github.com/middyjs/middy/blob/3dea624830f7f9fb2673657f113a1c72f1c5d28f/packages/ssm/index.js#L164
- AWS signature valid for 15 minutes, so if you have traffic and your lambda keeps running without cold starts you reuse your client
- eventually, the signature expires and your client starts throwing InvalidSignatureException
- what's interesting is that this line https://github.com/middyjs/middy/blob/3dea624830f7f9fb2673657f113a1c72f1c5d28f/packages/ssm/index.js#L125 sets undefined in the cache, and while it throwing an error all future invocations receiving undefined, which breaks completely everything

**Comment by willfarrell** [comment]
After some digging, I think I have a theory.

1. Middy is setup with an expiry
1. first Request is made, ssm is fetched, it get cached, timer is set to refresh the cache
1. While the lambda has been idle for a while, the timer expires triggering the refresh. A fetch promise is create but stopped somehow by AWS (the theory part)
1. next request comes in >5min later, the fetch promise w/ a now expire signature fails.

How to fix, All middlewares that fetch from aws services will need to catch `InvalidSignatureException` and force a retry during the request. I'll have to think on how to best  implement this.

Would love to hear if the above steps makes sense those running into this issue.

Ref:
https://repost.aws/knowledge-center/lambda-sdk-signature

**Comment by willfarrell** [comment]
I pushed a PR, if someone could test irl that would be great.

**Comment by HumbleBeck** [comment]
I'll try to copy-paste your changes, we are still on v4, and run it for a few days.

**Comment by willfarrell** [comment]
@HumbleBeck Any feedback on this?

**Comment by HumbleBeck** [comment]
Hi @willfarrell. While this bug rarely happens to us, I can confirm that the fix works, and it started recovering failed expired signature calls.

**Comment by willfarrell** [comment]
Awesome, I'll update the PR to cover all AWS service middleware (just in case) and merge in. Thanks a lot for testing it out.

**Comment by pranav-chefman** [comment]
Hi, you are still on version 4. One workaround is overriding the retry strategy and passing it to `middy`. 

```
const middy = require('@middy/core');
const ssm = require('@middy/ssm');
const {ConfiguredRetryStrategy} = require('@smithy/util-retry');


class ClockSkewRetryStrategy extends ConfiguredRetryStrategy {
  constructor(maxAttempts, computeNextBackoffDelay) {
    super(maxAttempts, computeNextBackoffDelay);
  }

  isRetryableError(errorType) {
    return errorType === 'CLIENT_ERROR' || super.isRetryableError(errorType);
  }
}

...

middy()
      .use(
        ssm({
          ...
          awsClientOptions: {
            retryStrategy: new ClockSkewRetryStrategy(3, 500),
          },
          ...
        })
      )
      .before(async (request) => {
        ...
      });
```

**Comment by Abdullah-Sajjad026** [comment]
Hey, I recently started facing this issue and I am on version 4. @pranav-chefman Have you tested the RetryStrategy fix and did it work for you?

**Comment by pranav-chefman** [comment]
@Abdullah-Sajjad026 yes

**Comment by Abdullah-Sajjad026** [comment]
@willfarrell — we're hitting this exact `InvalidSignatureException` cache-poisoning issue in production on `@middy/ssm@4.7.0` (the rejected `GetParameters` promise gets cached for the container's life, so a single stale-signed request during a long init→first-invoke gap poisons every subsequent request — for us it surfaced as `jwtSecret` resolving to `undefined` and ~1,100 auth failures from one warm container).

We can't move to v5 yet (we're CJS + serverless-esbuild and the ESM-only migration is a larger project), so as a stopgap we've backported your fix from #1150 onto our pinned 4.7.0 via patch-package-style patches:

- `@middy/util`: added `catchInvalidSignatureException` exactly as in #1150
- `@middy/ssm`: wrapped both `GetParametersCommand` and `GetParametersByPathCommand` sends with `.catch((e) => catchInvalidSignatureException(e, client, command))`

We've also paired it with `disablePrefetch: true` (so the fetch happens at invocation time rather than module load) and `cacheExpiry: -1`.

Could you sanity-check whether backporting #1150 to v4 this way is sound for this failure mode — specifically:

1. Is `e.__type === 'InvalidSignatureException'` the only error shape worth catching here, or should it also cover `RequestExpired` / cases where only `e.name` is set?
2. Is a single immediate re-send (no backoff) sufficient in practice, or did you consider more than one retry?
3. Anything about the cached-promise lifecycle in v4 that would make this backport behave differently than it does in v5?

Want to make sure we're not missing something before relying on this in prod. Thanks for the fix and the earlier analysis in this thread.
